synapse/storage/database.py

```diff
@@ -2431,6 +2431,65 @@ def make_in_list_sql_clause(
         return "%s IN (%s)" % (column, ",".join("?" for _ in iterable)), list(iterable)
 
 
+# These overloads ensure that `columns` and `iterable` values have the same length.
+@overload
+def make_tuple_in_list_sql_clause(
+    database_engine: BaseDatabaseEngine,
+    columns: Tuple[str, str],
+    iterable: Collection[Tuple[Any, Any]],
+) -> Tuple[str, list]:
+    ...
+
+
+def make_tuple_in_list_sql_clause(
+    database_engine: BaseDatabaseEngine,
+    columns: Tuple[str, ...],
+    iterable: Collection[Tuple[Any, ...]],
+) -> Tuple[str, list]:
+    """Returns an SQL clause that checks the given tuple of columns is in the iterable.
+
+    Args:
+        database_engine
+        columns: Names of the columns in the tuple.
+        iterable: The tuples to check the columns against.
+
+    Returns:
+        A tuple of SQL query and the args
+    """
+    if len(columns) == 0:
+        # Should be unreachable due to mypy, as long as the overloads are set up right.
+        if () in iterable:
+            return "TRUE", []
+        else:
+            return "FALSE", []
+
+    if len(columns) == 1:
+        # Use `= ANY(?)` on postgres.
+        return make_in_list_sql_clause(
+            database_engine, next(iter(columns)), [values[0] for values in iterable]
+        )
+
+    # There are multiple columns. Avoid using an `= ANY(?)` clause on postgres, as
+    # indices are not used when there are multiple columns. Instead, use an `IN`
+    # expression.
+    #
+    # `IN ((?, ...), ...)` with tuples is supported by postgres only, whereas
+    # `IN (VALUES (?, ...), ...)` is supported by both sqlite and postgres.
+    # Thus, the latter is chosen.
```

Comment:
The query plan for `IN (VALUES (?, ...), ...)` looks like
```
                                                                               QUERY PLAN
------------------------------------------------------------------------------------------------------------------------------------------------------------------------
 Nested Loop  (cost=5.67..2480429.48 rows=56325687 width=504) (actual time=1.669..1.671 rows=0 loops=1)
   ->  Nested Loop  (cost=5.11..923.05 rows=2657681 width=535) (actual time=1.668..1.669 rows=0 loops=1)
         ->  HashAggregate  (cost=4.55..6.55 rows=200 width=64) (actual time=0.321..0.418 rows=260 loops=1)
               Group Key: "*VALUES*".column1, "*VALUES*".column2
               ->  Values Scan on "*VALUES*"  (cost=0.00..3.25 rows=260 width=64) (actual time=0.004..0.158 rows=260 loops=1)
         ->  Index Scan using e2e_device_keys_json_uniqueness on e2e_device_keys_json k  (cost=0.56..4.58 rows=1 width=471) (actual time=0.004..0.004 rows=0 loops=260)
               Index Cond: ((user_id = "*VALUES*".column1) AND (device_id = "*VALUES*".column2))
   ->  Index Scan using device_uniqueness on devices d  (cost=0.56..0.93 rows=1 width=67) (never executed)
         Index Cond: ((user_id = k.user_id) AND (device_id = k.device_id))
         Filter: (NOT hidden)
 Planning time: 3.435 ms
 Execution time: 1.748 ms
(12 rows)
```
When `VALUES` is omitted, the query plan looks like:
```
 Nested Loop  (cost=670.12..680.76 rows=5 width=504) (actual time=0.537..0.566 rows=0 loops=1)
   ->  Bitmap Heap Scan on devices d  (cost=669.56..676.18 rows=1 width=67) (actual time=0.536..0.565 rows=0 loops=1)
         Recheck Cond: (((user_id = 'a01'::text) AND (device_id = 'b'::text)) OR ((user_id = 'c01'::text) AND ...
         Filter: (NOT hidden)
         ->  BitmapOr  (cost=669.56..669.56 rows=2 width=0) (actual time=0.534..0.563 rows=0 loops=1)
               ->  Bitmap Index Scan on device_uniqueness  (cost=0.00..2.57 rows=1 width=0) (actual time=0.008..0.008 rows=0 loops=1)
                     Index Cond: ((user_id = 'a01'::text) AND (device_id = 'b'::text))
               ->  Bitmap Index Scan on device_uniqueness  (cost=0.00..2.57 rows=1 width=0) (actual time=0.002..0.002 rows=0 loops=1)
                     Index Cond: ((user_id = 'c01'::text) AND (device_id = 'd'::text))
               ->  Bitmap Index Scan on device_uniqueness  (cost=0.00..2.57 rows=1 width=0) (actual time=0.002..0.002 rows=0 loops=1)
                     Index Cond: ((user_id = 'e01'::text) AND (device_id = 'f'::text))
               ->  ...
   ->  Index Scan using e2e_device_keys_json_uniqueness on e2e_device_keys_json k  (cost=0.56..4.58 rows=1 width=471) (never executed)
         Index Cond: ((user_id = d.user_id) AND (device_id = d.device_id))
 Planning time: 12.403 ms
 Execution time: 3.989 ms
(529 rows)
```

Reply:
AAAAAAAAARGH WHYYYYYYYYYYYYYYYYYYYYYYYY

Is this also true on pg14? :(

Reply:
I haven't got access to a postgres 14 db with realistic data to test with.
It only happens with tuples by the way. A regular `IN (...)` with single values gets turned into `=ANY`.